sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala

```diff
@@ -544,187 +392,56 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
     val location = visitLocationSpecList(ctx.locationSpec())
     // rowStorage used to determine CatalogStorageFormat.serde and
     // CatalogStorageFormat.properties in STORED AS clause.
-    val rowStorage = ctx.rowFormat.asScala.headOption.map(visitRowFormat)
-      .getOrElse(CatalogStorageFormat.empty)
-    val fileFormat = ctx.createFileFormat.asScala.headOption.map(visitCreateFileFormat) match {
-      case Some(f) =>
-        if (provider.isDefined) {
-          throw new ParseException("'STORED AS hiveFormats' and 'USING provider' " +
+    val rowFormatSerdeInfo = ctx.rowFormat.asScala.map(visitRowFormat)
+    val fileFormatSerdeInfo = ctx.createFileFormat.asScala.map(visitCreateFileFormat)
+    val serdeInfo =
+      (fileFormatSerdeInfo ++ rowFormatSerdeInfo).reduceLeftOption((x, y) => x.merge(y))
+    if (provider.isDefined && serdeInfo.isDefined) {
+      val sd = serdeInfo.get
+      if (sd.storedAs.isDefined) {
```

Comment:
Let's be consistent with https://github.com/apache/spark/pull/28026/files#diff-26fe6b511f4b8aedf5bdffae0d7c03aba69463c09fa4f881d77520125796e0faR2999